aesd-char-driver/main.c

```diff
@@ -0,0 +1,304 @@
+struct aesd_dev aesd_device;
+
+int aesd_open(struct inode *inode, struct file *filp)
+{
+    // Reference: Linux Device Drivers book
+
+    // Device information
+    struct aesd_dev *dev;
+
+    PDEBUG("open");
+    /**
+     * TODO: handle open
+     */
+
+    dev = container_of(inode->i_cdev , struct aesd_dev , cdev);
+
+    filp->private_data = dev;
+
+    return 0;
+}
+
+int aesd_release(struct inode *inode, struct file *filp)
+{
+    PDEBUG("release");
+    /**
+     * TODO: handle release
+     */
+
+    // Reference: Linux Device Drivers book
+    // No code required as device has no hardware to shut down
+    return 0;
+}
+
+ssize_t aesd_read(struct file *filp, char __user *buf, size_t count,
+                loff_t *f_pos)
+{
+    ssize_t retval = 0;
+
+    // Reference: Linux Device Drivers book
+    // Store the filp paramter in the user structure
+    struct aesd_dev *dev = filp->private_data;
+    size_t entry_byte_offs = 0;
+    ssize_t res_bytes = 0;
+    struct aesd_buffer_entry * ret_entry_buffer;
+    PDEBUG("read %zu bytes with offset %lld",count,*f_pos);
+    /**
+     * TODO: handle read
+     */
+
+    // Acquire the mutex lock
+    mutex_lock(&aesd_device.rw_mutex_lock);
+
+    ret_entry_buffer = aesd_circular_buffer_find_entry_offset_for_fpos(&dev->rw_circular_buffer , *f_pos , &entry_byte_offs);
+
+    // Error check
+    if (ret_entry_buffer == NULL){
+        *f_pos = 0;
+        goto clean;
+    }
+
+    // If the difference between size of the data entry is positive but less than count, part of data has been transferred
+    if ((ret_entry_buffer->size - entry_byte_offs) < count){
+        *f_pos += (ret_entry_buffer->size - entry_byte_offs);
+        res_bytes = ret_entry_buffer->size - entry_byte_offs;
+
+    } 
+    // Optimal case: All bytes can be transferred
+    else{
+        *f_pos += count;
+        res_bytes = count;
+    }
+
+    if (copy_to_user(buf, (ret_entry_buffer->buffptr + entry_byte_offs), res_bytes)) {
+		retval = -EFAULT;
+		goto clean;
+	}
+
+    retval = res_bytes;
+
+    clean : mutex_unlock(&aesd_device.rw_mutex_lock);
+
+    return retval;
+}
+
+ssize_t aesd_write(struct file *filp, const char __user *buf, size_t count,
+                loff_t *f_pos)
+{
+    ssize_t retval = 0;
+    // Write buffer
+    char *write_buffer; 
+
+    // Flag indicating complete packet
+    bool is_packet_complete = false;
+
+    // Stores the total length of the packet
+    int total_packet_len = 0;
+
+    // Holds the value to be added to the circular buffer
+    struct aesd_buffer_entry write_cb_entrys; 
+    struct aesd_dev *dev;
+
+    // Iterates over incoming data while looking for '\n'
+    int i; 
+
+    int realloc_size = 0; 
+
+    char *ret_ptr;
+
+    PDEBUG("write %zu bytes with offset %lld",count,*f_pos);
+    /**
+     * TODO: handle write
+     */
+
+
+    PDEBUG("write %zu bytes with offset %lld",count,*f_pos);
+
+    dev = filp->private_data;
+
+    mutex_lock(&aesd_device.rw_mutex_lock);
```

Comment:
@KhyatiSatta is this the minimum locking interval?  What is `aesd_device.rw_mutex_lock` locking? See discussion in class today.